zebra-rpc/src/sync.rs

```diff
@@ -0,0 +1,385 @@
+
+                // If the block hash matches the output from the `getbestblockhash` RPC method, we can wait until
+                // the best block hash changes to get the next block.
+                if block_hash == target_tip_hash {
+                    break false;
+                }
+            };
+
+            if should_reset_non_finalized_state {
+                self.try_catch_up_with_primary().await;
+                let block = self.finalized_chain_tip_block().await.expect(
+                    "should have genesis block after successful bestblockheightandhash response",
+                );
+
+                last_chain_tip_hash = block.hash;
+                self.non_finalized_state =
+                    NonFinalizedState::new(&self.non_finalized_state.network);
+                self.update_channels(block);
+            }
+        }
+    }
+
+    /// Tries to catch up to the primary db instance for an up-to-date view of finalized blocks.
+    // TODO: Use getblock RPC if it fails to catch up to primary?
+    async fn try_catch_up_with_primary(&self) {
+        let db = self.db.clone();
+        tokio::task::spawn_blocking(move || {
+            if let Err(catch_up_error) = db.try_catch_up_with_primary() {
+                tracing::warn!(?catch_up_error, "failed to catch up to primary");
+            }
+        })
+        .wait_for_panics()
+        .await
+    }
+
+    /// If the non-finalized state is empty, tries to catch up to the primary db instance for
+    /// an up-to-date view of finalized blocks.
+    ///
+    /// Returns true if the non-finalized state is empty and the target hash is in the finalized state.
+    async fn is_finalized_tip_change(&self, target_tip_hash: block::Hash) -> bool {
+        self.non_finalized_state.chain_count() == 0 && {
+            let db = self.db.clone();
+            tokio::task::spawn_blocking(move || {
+                if let Err(catch_up_error) = db.try_catch_up_with_primary() {
+                    tracing::warn!(?catch_up_error, "failed to catch up to primary");
+                }
+                db.contains_hash(target_tip_hash)
+            })
+            .wait_for_panics()
+            .await
+        }
+    }
+
+    /// Returns the current tip height and hash
+    async fn next_block_height_and_prev_hash(&self) -> (block::Height, block::Hash) {
+        if let Some(tip) = self.non_finalized_state.best_tip() {
+            Some(tip)
+        } else {
+            let db = self.db.clone();
+            tokio::task::spawn_blocking(move || db.tip())
+                .wait_for_panics()
+                .await
+        }
+        .map(|(current_tip_height, current_tip_hash)| {
+            (
+                current_tip_height.next().expect("should be valid height"),
+                current_tip_hash,
+            )
+        })
+        .unwrap_or((Height::MIN, GENESIS_PREVIOUS_BLOCK_HASH))
+    }
+
+    async fn finalized_chain_tip_block(&self) -> Option<ChainTipBlock> {
```

Comment:
```suggestion
    /// Reads the finalized tip block from the secondary db instance and converts it to a [`ChainTipBlock`].
    async fn finalized_chain_tip_block(&self) -> Option<ChainTipBlock> {
```

Reply:
Applied in https://github.com/ZcashFoundation/zebra/pull/8596/commits/15feed173daced582340fdcdb4d4bb8cc9593a39.